cmd/auth-provider-gcp/app/getcredentials.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/spf13/cobra"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/cloud-provider-gcp/cmd/auth-provider-gcp/credentialconfig"
+	"k8s.io/cloud-provider-gcp/cmd/auth-provider-gcp/plugin"
+	klog "k8s.io/klog/v2"
+	"net/http"
+)
+
+var (
+	authFlow string
+)
+
+const (
+	gcrAuthFlow             = "gcr"
+	dockerConfigAuthFlow    = "dockercfg"
+	dockerConfigURLAuthFlow = "dockercfg-url"
+)
+
+// NewGetCredentialsCommand returns a cobra command that retrieves auth credentials after validating flags.
+func NewGetCredentialsCommand() (*cobra.Command, error) {
+	cmd := &cobra.Command{
+		Use:   "get-credentials",
+		Short: "Get authentication credentials",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			klog.Infof("get-credentials %s", authFlow)
+			transport := utilnet.SetTransportDefaults(&http.Transport{})
+			var provider credentialconfig.DockerConfigProvider
+			if authFlow == gcrAuthFlow {
+				provider = plugin.MakeRegistryProvider(transport)
+			} else if authFlow == dockerConfigAuthFlow {
+				provider = plugin.MakeDockerConfigProvider(transport)
+			} else {
+				provider = plugin.MakeDockerConfigURLProvider(transport)
+			}
+			authCredentials, err := plugin.GetResponse(provider)
+			if err != nil {
+				return err
```

Comment:
Have we tried triggering an error case and verified how its handled by Kubelet?

Reply:
Error handling is a little tricky at the moment since the existing credential provider code doesn't give any indication that an error has actually occurred outside of checking the log files, as it simply returns an empty set of credentials. (This behavior is currently mirrored on the kubelet side of the plugin framework - see https://github.com/kubernetes/kubernetes/blob/1ea29cec9c923f18774092df9c3b7d0de7a193c2/pkg/credentialprovider/plugin/plugin.go#L199.) `err` here will only be non-nil if there's a problem parsing the time duration in the `KUBE_SIDECAR_CACHE_DURATION` environment variable. It might be possible to watch for an empty credential set and return an error there, but that seems like an assumption that might hold true in all cases.

Now, more generally speaking (i.e, outside of this particular `if` statement), I've watched how kubelet handles errors - if the plugin returns a non-zero error code, kubelet will continue to run, but no credentials will be fetched, so eventually any image retrieval will fail when kubelet attempts to search the keyring for the proper username/password pair.